Make module and email translations work again when using a theme other than classic

eternoendless opened the pull request:
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | 1.7.6.x
| Description?  | Fixes a bug where it would be impossible to translate modules and email subjects when the current theme was anything other than "classic"
| Type?         | bug fix
| Category?     | BO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | Fixes #15133
| How to test?  | Install a different theme than classic. Test all the translation types (except Email body which is handled differently and not affected by this change).

## More information

There was much confusion between the "selected" element (a module name when translating modules, a theme name when translating themes) and the selected theme. Up until now, it was half-managed by request parameters, half-forced to the current theme.

This change clarifies things a little:

- Module translations are no longer attached to the currently selected theme (you'll have to edit the theme translations if you want a different translation for each theme; this requires the specific wording to be used in the theme's templates or else it won't appear in the translation interface)
- Only theme translations are attached to a theme. All other translations are now global.
- Email translations no longer are attached to an invalid theme called "subject"

Regarding #14809, we'll have to discuss changes in how Mail translations are managed so that email translations work like modules (where `type` would be 'mails' and `selected` would be either `subject` or `body`). But this will be problematic, because I think that in difference to email subjects, which are global, mail bodies are theme-dependent.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/prestashop/prestashop/15139)
<!-- Reviewable:end -->

Commit:
Fix translations interface

- Module translations are no longer attached to the currently selected theme
- Email translations no longer are attached to an invalid theme called "subject"

Commit:
Build assets

PierreRambaud commented on src/PrestaShopBundle/Resources/views/Admin/Translations/overview.html.twig:
```diff
@@ -35,7 +35,9 @@
             locale: '{{ app.request.query.get('locale') }}',
-            selected: '{{ 'modules' == app.request.query.get('type') ? theme.name : app.request.query.get('selected') }}',
+            type: '{{ app.request.query.get('type') }}',
+            currentTheme: {{ theme.name|json_encode|raw }},
```
watchout with raw.
```suggestion
            currentTheme: '{{ theme.name|escape('js') }}',
```

eternoendless replied:
That won't work with null, functions are not json encodable, and html would be encapulated in a string. I'm I missing something?

PierreRambaud replied:
If it's not working with null.
```suggestion
            currentTheme: {% if theme.name === null %}null{% else %}'{{ theme.name|escape('js') }}'{% endif %},
```

eternoendless replied:
Looks overcomplicated when json_encode is build exactly for that purpose: it outputs the right Javascript syntax for whatever value you throw at it.

mickaelandrieu commented:
> Module translations are no longer attached to the currently selected theme (you'll have to edit the theme translations if you want a different translation for each theme; this requires the specific wording to be used in the theme's templates or else it won't appear in the translation interface)

If I've well understood, this introduces is a regression to fix a bug, right?

matks commented on admin-dev/themes/new-theme/js/app/pages/translations/components/principal/index.vue:
```diff
@@ -129,21 +129,22 @@
-          totalMissingTranslationsString = this.trans('label_missing').replace('%d', this.currentDomainTotalMissingTranslations);
+        if (this.currentDomainTotalMissingTranslations) {
+          if (this.currentDomainTotalMissingTranslations === 1) {
+            totalMissingTranslationsString = this.trans('label_missing_singular');
```
Some extra polishing 😛

eternoendless replied:
A little bit of the ol' boy scout rule

matks commented on src/PrestaShopBundle/Controller/Api/TranslationController.php:
```diff
@@ -140,10 +140,24 @@ public function listTreeAction(Request $request)
+                    $tree = $this->getMailsSubjectTree($lang, $search);
+                    break;
+
+                default:
```
What else can be translated ?
If we expect nothing else to be translated, maybe we should throw an exception in the `default` ?

matks replied:
(maybe not in a patch version though 😛)

jolelievre replied:
Yes there are other types, like `back` `others`

eternoendless replied:
^ What jo answered

matks commented on src/PrestaShopBundle/Controller/Api/TranslationController.php:
```diff
@@ -307,65 +321,79 @@ protected function guardAgainstInvalidTranslationResetRequest($content)
      * @return array
      */
-    private function getNormalTree($lang, $type, $selected, $search = null)
+    private function getNormalTree($lang, $type, $theme, $search = null)
```
If `getNormalTree` is supposed to only be called for Themes, I suggest we strongly validate the input by checking `$theme` actually exists and refers to an installed theme, or throw an exception.

matks replied:
(maybe not in a patch version though 😛)

jolelievre replied:
Since we planned a refacto maybe it should be done in the refacto?

eternoendless replied:
This is called on line 159 with a null theme (`BO` translations and `other`)

matks commented on src/PrestaShopBundle/Controller/Api/TranslationController.php:
```diff
@@ -307,65 +321,79 @@ protected function guardAgainstInvalidTranslationResetRequest($content)
      * @return array
      */
-    private function getModulesTree($lang, $type, $selected, $search = null)
+    private function getModulesTree($lang, $selectedModuleName, $search = null)
```
I suggest same defensive logic: if `$selectedModuleName` is illogical, we crash. Translations are a mess right now, this would clear the path and avoid ambiguous successfull behaviors

matks replied:
(maybe not in a patch version though 😛)

eternoendless replied:
I think it will crash but deeper in the call stack. But we agree that this whole thing needs to be refactored.

matks reviewed (commented):
OK for me, same feedback than @PierreRambaud

jolelievre commented on src/PrestaShopBundle/Controller/Api/TranslationController.php:
```diff
@@ -307,65 +321,79 @@ protected function guardAgainstInvalidTranslationResetRequest($content)
+        $moduleProvider->setModuleName($selectedModuleName);
 
-        $treeBuilder = new TreeBuilder($locale, $this->getSelectedTheme());
+        $treeBuilder = new TreeBuilder($locale, $theme);
```
So `$theme` is always `null` ?

eternoendless replied:
Yes when translating modules. Otherwise you'd have to translate your modules again for every theme (including if you switch themes). With this change only the wordings used in module templates overridden by a theme will be allowed to have a specific translation for that theme. This could be improved later by adding an option that brings ALL wordings, including core and modules, and that allows translating them for a specific theme.

jolelievre replied:
then you don't need this `$theme` variable you can simply call `$treeBuilder = new TreeBuilder($locale, null);`

eternoendless replied:
I did it so that the parameter role is evident, but it could be removed as well later.

jolelievre commented on src/PrestaShopBundle/Controller/Api/TranslationController.php:
```diff
@@ -307,65 +321,79 @@ protected function guardAgainstInvalidTranslationResetRequest($content)
+     * @return string
      */
-    private function getSelectedTheme()
+    private function getCurrentThemeName()
```
So it seems this method is no longer used?

eternoendless replied:
Right, I'll delete it

matks commented:
Restarting Travis again

matks commented:
And again

matks commented:
And again

I could _almost_ believe the test fails for a good reason

PierreRambaud commented:
And again!

Always the same random reason :/

Commit:
Deleted unused private method

eternoendless commented:
I removed the unused method and rebased.

> If I've well understood, this introduces is a regression to fix a bug, right?

Yes and no. It made no sense, it was counterintuitive, and in the end, it didn't work. It's a regression in the same way as fixing a bug is a regression: "But I liked that bug, put it back!" 😛

The system has to be rethought, and in most importantly... **specified**.

matks commented:
Thank you @eternoendless

sathishkumar16 commented:
When we will get this update in the prestashop

eternoendless commented:
@sathishkumar16 Soon.

sathishkumar16 commented:
@eternoendless  How many days will it take to get the updates officially, because we are planning to production within a month for us the translation module is very important before going to prod.

eternoendless commented:
@sathishkumar16 The QA process is looking good so it won't be long, probably within the week. You can follow the discussion on [Gitter](https://gitter.im/PrestaShop/General) or wait for the announcement on [Build](http://build.prestashop.com/).
